satpy/readers/fci_l1c_fdhsi.py

```diff
@@ -291,44 +329,30 @@ def _ir_calibrate(self, radiance, measured, root):
                         v.attrs.get("long_name",
                                     "at least one necessary coefficient"),
                         root))
-                return xr.DataArray(
-                        da.full(shape=radiance.shape,
-                                chunks=radiance.chunks,
-                                fill_value=np.nan),
-                        dims=radiance.dims,
-                        coords=radiance.coords,
-                        attrs=radiance.attrs)
-
-        Lv = radiance * coef
-        vc = 1e6/wl_c  # from wl in um to wn in m^-1
+                return radiance*np.nan
+
         nom = c2 * vc
-        denom = a * np.log(1 + (c1 * vc**3) / Lv)
+        denom = a * np.log(1 + (c1 * vc**3) / radiance)
 
         res = nom / denom - b / a
         res.attrs["units"] = "K"
         return res
 
-    def _vis_calibrate(self, radiance, measured):
+    def _vis_calibrate(self, radiance, key):
         """VIS channel calibration."""
-        # radiance to reflectance taken as in mipp/xrit/MSG.py
-        # again FCI User Guide is not clear on how to do this
+        measured, _ = self.get_channel_dataset(key.name)
+
+        cesi = self[measured + "/channel_effective_solar_irradiance"]
 
-        cesilab = measured + "/channel_effective_solar_irradiance"
-        cesi = self[cesilab]
         if cesi == cesi.attrs.get(
                 "FillValue", default_fillvals.get(cesi.dtype.str[1:])):
             logger.error(
                 "channel effective solar irradiance set to fill value, "
                 "cannot produce reflectance for {:s}.".format(measured))
-            return xr.DataArray(
-                    da.full(shape=radiance.shape,
-                            chunks=radiance.chunks,
-                            fill_value=np.nan),
-                    dims=radiance.dims,
-                    coords=radiance.coords,
-                    attrs=radiance.attrs)
-
-        sirr = float(cesi)
-        res = radiance / sirr * 100
+            return radiance*np.nan
+
+        sun_earth_distance = np.mean(self["state/celestial/earth_sun_distance"]) / 149597870.7  # [AU]
```

Comment:
I see the earth-sun distance have 6000 values in a segment, what do they correspond to ? Just curiosity, I understand the mean is good enough here.

Reply:
They're the earth-sun distance values during the repeat cycle with a 0.1 seconds time step :) Using the `index_map`, you can map to each image pixel the corresponding value, valid at the exact pixel acquisition time. There are a number of geometric parameters that work like this in the product (`data.swath_direction, data.swath_number, time (of acquisition), state.platform.subsatellite_latitude, state.platform.subsatellite_longitude, state.platform.platform_altitude (of the satellite), state.celestial.subsolar_latitude, state.celestial.subsolar_longitude, state.celestial.earth_sun_distance, and state.celestial.sun_satellite_distance`).

Reply:
Thanks for the clarification !